packages/runtime/src/runtime.ts

```diff
@@ -324,19 +347,160 @@ export class Runtime {
     account.optInToASA(assetIndex, assetHolding);
   }
 
+  // creates asset configuration transaction object.
+  mkAssetConfigTx (sender: string, assetId: number, fields: AssetModFields, flags: TxParams): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetConfigTxnWithSuggestedParams(
+      sender,
+      encodeNote(flags.note, flags.noteb64),
+      assetId,
+      fields.manager !== "" ? fields.manager : undefined,
+      fields.reserve !== "" ? fields.reserve : undefined,
+      fields.freeze !== "" ? fields.freeze : undefined,
+      fields.clawback !== "" ? fields.clawback : undefined,
+      mockSuggestedParams(flags, this.round),
+      false
+    );
+    const asset = this.getAssetDef(assetId);
+    if (asset.manager !== sender) {
+      throw new Error("Only Manager account can modify asset");
+    }
+  }
+
   /**
-   * Returns Asset Definitions
+   * https://developer.algorand.org/docs/features/asa/#modifying-an-asset
+   * Modifies asset fields
+   * @param sender sender address
    * @param assetId Asset Index
+   * @param fields Asset modifying fields
+   * @param payFlags Transaction Parameters
    */
-  getAssetDef (assetId: number): AssetDef {
-    const addr = this.store.assetDefs.get(assetId);
-    if (addr === undefined) {
-      throw new TealError(ERRORS.ASA.ASSET_NOT_FOUND, { assetId: assetId });
+  modifyAsset (sender: string, assetId: number, fields: AssetModFields, payFlags: TxParams): void {
+    this.mkAssetConfigTx(sender, assetId, fields, payFlags);
+    const creatorAcc = this.getAssetAccount(assetId);
+    creatorAcc.modifyAsset(assetId, fields);
+  }
+
+  // creates asset freeze transaction object.
+  mkAssetFreezeTx (
+    sender: string, assetId: number, target: string, state: boolean, flags: TxParams): void {
+    // this funtion is called only for validation of parameters passed
+    algosdk.makeAssetFreezeTxnWithSuggestedParams(
```

Comment:
Is it possible to have these types in `ExecParams` itself (Since we already have `AssetTransfer`) ? Then i think we could make these transactions at a single place (in mkTransaction) instead of in different functions. Also, it would be common for `algob` & `runtime` as well. Can we do that?

Reply:
seeing from user perspective i think using functions for ASA is easier because ExecParams already have application types + payment + lsig, not sure though if we should change current structure

Reply:
I think from the user perspective, keeping all execution transactions in `TransactionType` makes more sense, user can use `TransactionType.ModifyAsset`, `TransactionType.TransferAsset`,.. instead of looking for each specific function.

Also think about group transactions, by this structure we can't support asset `freeze`, `clawback`,.. in a group of transactions.

Reply:
shouldn't we do this for opt-in for asa or app as well ? also for update, app ?
we should be consistent then.

what if group transaction include opt-in + payment ?

Reply:
For update app: yes, i agree. 

For createApp, opt-in i think these are used for setting up (deploying) the application/asset. In examples we use them during the deployment mode (not while RUN, where we actually use `executeTransaction`). Also couldn't find any examples in reference where creation/opt-in transactions are used in a group, but that doesn't mean it's not possible.

Anyway, for asset freeze, clawback,.. transactions, we should keep them in `ExecParams` because they are used in the RUN mode, they are not related to setting up of an asset. (User will use them in `/scripts/transfer` in examples probably).

I was planning to add this example: [https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/](https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/), but we require asset clawback tx in a group here. What do you think?

Reply:
I think create should in transaction type as well, because group transaction create + pay.

Reply:
Yes, possible. But please update the asset transactions here, this way we would be moving in the right direction, not the opposite.

Reply:
agree for asset freeze, clawback etc... but for create and opt-in we should keep them in ExecParams too because that's the right way,
also if a user is making an application using algob, what if there is a transaction with create and pay or opt-in and pay
or pay and opt-in, for ex: you can only use our application if you pay some amount.
do you think we should avoid that ? @ratik21

Reply:
Currently we are avoiding it, but we shouldn't avoid that. Continued on discord.